controller/src/main/java/io/pravega/controller/store/stream/ZKGarbageCollector.java

```diff
@@ -0,0 +1,238 @@
+    // dedicated thread executor for GC which will not interfere with rest of the processing.
+    private final ScheduledExecutorService gcExecutor;
+    private final long periodInMillis;
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final CompletableFuture<Void> acquiredLeadership = new CompletableFuture<>();
+    
+    ZKGarbageCollector(String gcName, ZKStoreHelper zkStoreHelper, Supplier<CompletableFuture<Void>> gcProcessing, Duration gcPeriod) {
+        Preconditions.checkNotNull(zkStoreHelper);
+        Preconditions.checkNotNull(gcProcessing);
+        Preconditions.checkArgument(gcPeriod != null && !gcPeriod.isNegative());
+
+        this.currentBatch = new AtomicLong();
+        this.gcName = gcName;
+        this.leaderPath = String.format(LEADER_PATH, gcName);
+        this.batchPath = String.format(BATCH_PATH, gcName);
+        this.zkStoreHelper = zkStoreHelper;
+        this.gcProcessing = gcProcessing;
+        this.periodInMillis = gcPeriod.toMillis();
+        this.gcExecutor = Executors.newSingleThreadScheduledExecutor();
+    }
+
+    @Override
+    protected void doStart() {
+        long quantized = computeNewBatch();
+        byte[] b = getBytes(quantized);
+        // 1. create znode for storing latest batch id. If the batch id exists, get the value from the store.
+        // We will later register watch on the path and keep receiving any changes to its value.
+        zkStoreHelper.createZNodeIfNotExist(batchPath, b)
+                .thenCompose(x -> zkStoreHelper.getData(batchPath))
+                .thenAccept(data -> {
+                    currentBatch.set(BitConverter.readLong(data.getData(), 0));
+                })
+                .thenRun(this::initialize)
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStarted();
+                    }
+                    latch.complete(null);
+                });
+    }
+
+    @Override
+    protected void doStop() {
+        latch.thenAccept(v -> {
+            if (gcLoop != null) {
+                gcLoop.cancel(true);
+                gcLoop.whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStopped();
+                    }
+                });
+            } else {
+                notifyStopped();
+            }
+        });
+    }
+
+    @Override
+    public Long getLatestBatch() {
+        return currentBatch.get();
+    }
+
+    @SneakyThrows
+    private void initialize() {
+        // 1. register watch on batch path
+        watch = registerWatch(batchPath);
+
+        // 2. attempt to acquire leadership
+        leaderLatch = electLeader(leaderPath, acquiredLeadership);
+
+        // 3. if this acquires leadership, then schedule periodic garbage collection.
```

Comment:
Having only one node responsible for garbage collection seems a serious limitation of this approach. Is there a way a way we can have the work distributed?